### PR TITLE
Stream setup

### PR DESCRIPTION
**Pull request by kgopal492**
Wrote most of code for hashtag functionality. Some testing needs to be done.

### CODE REVIEW

File: cpp/client.cc [hunk 1]
```diff
@@ -98,3 +98,21 @@ void Client::monitor(std::string username,
...
+
+void Client::stream(std::string hashtag,
+                     std::unique_ptr<chirp::ServiceLayer::Stub>& stub_) {
+  grpc::ClientContext context;
+  chirp::StreamRequest request;
+  request.set_hashtag(hashtag);
+  chirp::StreamReply reply;
+  std::unique_ptr<grpc::ClientReader<chirp::StreamReply> > stream_handle(
+      stub_->stream(&context, request));
+  while (true) {
```
Comment (barath) [inline cpp/client.cc:112]: You should describe the termination condition, since it's not apparent.

File: cpp/client.h [hunk 2]
```diff
@@ -22,6 +22,9 @@ class Client {
   // streams chirps from users username is following
   void monitor(std::string username,
                std::unique_ptr<chirp::ServiceLayer::Stub>& stub_);
+  // streams chirps made containing a provided hashtag
+  void stream(std::string hashtag,
```
Comment (barath) [inline cpp/client.h]: const reference?

File: cpp/service.cc [hunk 3]
```diff
@@ -100,8 +100,38 @@ chirp::Chirp ServiceLayer::chirp(const std::string& username,
     const std::string this_chirp_id = my_id;
     store_->put(chirp_parent_id, my_id);
   }
+
+  // TODO: check whether hashtag contains '#' character
+  std::string temp_text = text;
+  size_t hash_index = temp_text.find("#");
+  while (hash_index != std::string::npos) {
```
Comment (barath) [inline cpp/service.cc:113]: Break string handling into separate function?

File: cpp/service_layer_service_impl.cc [hunk 4]
```diff
@@ -94,3 +94,45 @@ grpc::Status ServiceLayerServiceImpl::monitor(
...
+  std::deque<chirp::Chirp> found_chirps;
+  while (keep_streaming) {
+    seconds = std::chrono::duration_cast<std::chrono::seconds>(
+        std::chrono::system_clock::now().time_since_epoch());
+    useconds = std::chrono::duration_cast<std::chrono::microseconds>(
+        std::chrono::system_clock::now().time_since_epoch());
+    found_chirps = service_.stream(request->hashtag(), initial_time);
+    initial_time.set_seconds(seconds.count());
+    initial_time.set_useconds(useconds.count());
+    for (chirp::Chirp c : found_chirps) {
```
Comment (barath) [inline cpp/service_layer_service_impl.cc]: const reference?

File: cpp/service_layer_service_impl.cc [hunk 5]
```diff
@@ -94,3 +94,45 @@ grpc::Status ServiceLayerServiceImpl::monitor(
   }
   return grpc::Status::OK;
 }
+
+grpc::Status ServiceLayerServiceImpl::stream(
+    grpc::ServerContext* context, const chirp::StreamRequest* request,
+    grpc::ServerWriter< ::chirp::StreamReply>* stream) {
+  chirp::Timestamp initial_time;
+  std::chrono::seconds seconds =
```
Comment (barath) [inline cpp/service_layer_service_impl.cc]: Maybe choose more descriptive variable names?

File: cpp/service_layer_service_impl.cc [hunk 6]
```diff
@@ -94,3 +94,45 @@ grpc::Status ServiceLayerServiceImpl::monitor(
...
+        reply.set_allocated_chirp(this_chirp);
+        chirp::StreamReply sendingReply = reply;
+        stream->Write(sendingReply);
+        read_chirps.insert(c.id());
+      }
+    }
+    if (context->IsCancelled()) {
+      keep_streaming = false;
+    }
+    usleep(20);
```
Comment (barath) [inline cpp/service_layer_service_impl.cc]: Generally better to not have a hardcoded sleep of any kind.

File: cpp/client.cc (hunk 1 above)
Comment (kgopal492) [reply cpp/client.cc:112]: added a comment

File: cpp/client.h (hunk 2 above)
Comment (kgopal492) [reply cpp/client.h]: changed

File: cpp/service.cc (hunk 3 above)
Comment (kgopal492) [reply cpp/service.cc:113]: added a helper function.

File: cpp/service_layer_service_impl.cc (hunk 4 above)
Comment (kgopal492) [reply cpp/service_layer_service_impl.cc]: changed

File: cpp/service_layer_service_impl.cc (hunk 6 above)
Comment (kgopal492) [reply cpp/service_layer_service_impl.cc]: removed.

File: cpp/service_layer_service_impl.cc (hunk 5 above)
Comment (kgopal492) [reply cpp/service_layer_service_impl.cc]: fixed

File: cpp/service.cc [hunk 7]
```diff
@@ -190,3 +225,27 @@ std::deque<chirp::Chirp> ServiceLayer::monitor(const std::string& username,
...
+  const std::deque<std::string>& hashtag_chirps =
+      store_->get(hashtag_key);
+  for (const std::string& id : hashtag_chirps) {
+    std::deque<std::string> this_chirps_values =
+        store_->get(kchirpValue_ + id);
+    chirp::Chirp thisChirp;
+    if (this_chirps_values.size() > 0) {
+      thisChirp.ParseFromString(this_chirps_values.at(0));
+      int64_t myMicroSeconds = thisChirp.timestamp().useconds();
+      int64_t baslineSeconds = start.useconds();
```
Comment (jillianKhoo) [inline cpp/service.cc]: is baslineSeconds ever used?

Comment (kgopal492) [reply cpp/service.cc]: You're right. I removed the unused variables in this function.

**Review by jillianKhoo — APPROVED** [review]

**Comment by kgopal492** [comment]
@barath approved?

**Review by barath — APPROVED** [review]

**Pull request merged**